Throw on code 306

jonchurch opened the pull request:
This is just a suggestion based on issue #18.

I have removed code 306 after `codes.json` is required to make `status(306)` throw as is indicated in the docs. 

I assume this would be a breaking change. 

The main question is:

### Should 306 throw an error? 

Assuming it should, when fetching the IANA codes we could remove it from the response before writing it to `iana.json`. Or it could be manually edited like I see has been done for some other codes like in c46c4d4d8b85814c0e65f99af8ce3137bca6b0ea.

If 306 should not throw, the docs can just be updated.

helio-frota commented:
@jonchurch makes sense :+1: 

```
status(306) // throws, as it's not supported by node.js
```

Currently it is not throwing and we are able to log the code to stdout: 

```
306
      ✓ should FOO
```

Not sure if it aggregates as extra info but I found this thread on IANA codes https://github.com/nodejs/node/pull/1470

To be clear: 
Indeed, 306 is not there https://github.com/nodejs/node/blob/master/lib/_http_server.js#L98-L99

Then if the idea is to throw, according the docs:  `throws, as it's not supported by node.js`, then your PR does it :+1:

dougwilson commented:
Yea, it looks like there are two issues here:

(1) docs issue: the throw is unrelated to if a code is "supported" by Node.js; it is just if a code is known to exist
(2) 306 existing: it shows up in iana because it _used_ to be registered, but was unregistered. The iana fetch code should filter out the (unused) code directly (vs what this pr is doing deleting after the fact).

jonchurch commented:
Updated the PR to filter `306` out at fetch time.

Note, IANA seems to require a `User-Agent` header to be present when making this request now. When running the fetch script unmodified I received the response `Status 403 Forbidden: User-Agent required. Contact iana@iana.org with questions.`.

I updated the fetch script to include `User-Agent: jshttp/statuses` header.

I also added a regression test to make sure code 306 throws an error.

dougwilson commented:
Ah, I was really curious why the User-Agent was added, thanks for the explaination 👍

dougwilson commented on scripts/fetch-iana.js:
```diff
@@ -19,7 +20,7 @@ https.get(URL, function onResponse (res) {
       var obj = row.reduce(reduceRows, {})
 
-      if (obj.description !== 'Unassigned') {
+      if (obj.description !== 'Unassigned' && obj.value !== '306') {
```
I'm kind of on the fence on if this should specifically blacklist _306_ code, or maybe just ones where the message is (unused) or similar. I was originally picturing in my mind the blacklist being against `(Unused)`, but of course there is no other code like it, so it wouldn't really make a difference. Since a reason phrase can be any obs-text, there is nothing wrong with it per-se, so I suppose a blacklist on the specific code is good enough.

jonchurch replied:
I'm on the fence too, but I think it serves as self documentation. It's a unique case, so handling it in code as unique seems appropriate. 

The logical question would be, do we want to remove any code that in the future is documented as `(Unused)`? 

I can change it right now if you'd prefer to see us filter out all `(Unused)`, otherwise I think singling out 306 is appropriate.

dougwilson replied:
No, sorry, it was just a comment :) not a request for a change. I just after that followed the IANA registration and I think there is no clear (good) way to make such a distinction, so a filter on the code seems appropriate.

dougwilson replied:
Since this does not need to be performant or anything and there are two filters, I flipped them so we could add a comment about why 306 is ignore for reference :)

dougwilson commented:
Hye @jonchurch I'm going to cherry-pick your user-agent change on to master prior to merging this PR; this is mainly because it's logically a separate change from this, though you happened to find it while working on this (great find!). I hope that is OK.

jonchurch commented:
No problem

dougwilson commented:
Cool. LMK if you have any issue with a couple nit fixes I pushed up here 👍 . I'm working to get this module release out as a 2.0 finally and bring this repo down to 0 issues / 0 PRs.

jonchurch commented:
LGTM

dougwilson commented:
Nice. Thanks for your work on this @jonchurch ; as we continue to push and clean up these repos, we will get closer and closer to a place where PRs are more likely to land quickly 🎉

dougwilson commented:
Changes squashed and landed.
